### PR TITLE
feat(Question Bank): Click question to add to student response input

### LINKED ISSUE #1413: feat(Question Bank): Click question to add to response input
- Allow students to click on questions in the question bank to add the question text into the response input. The student can then modify the question if they would like before they submit it.
- Allow authors to enable and disable this feature
- After a student uses a question, indicate that the question has been used
- Store which question was used in the student response
- When questions are authored, they should now have a unique id
- Allow authors to change the name of the question bank

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

- Allow students to click on questions in the question bank to add the question text into the response input. The student can then modify the question if they would like before they submit it.
- Allow authors to enable and disable this feature
- Allow authors to change the name of the question bank
- After a student uses a question, indicate that the question has been used
- Store which question was used in the student response
- When questions are authored, they should now have a unique id
  - New Question Bank content is marked as version 2 which contains question objects that have an id and text field
  - Old Question Bank content contains question strings

## Test

As a teacher
1. Create a new Peer Chat component
2. Enable the Question Bank in the Peer Chat
3. Set a Custom Label for the Question Bank
4. Enable Student Click To Add
5. Set a Reference Component
6. Set Grouping Logic
7. Add some Question Bank Rules

As a student
1. Go to the Peer Chat step and make sure the Question Bank shows up and the questions are clickable
2. Click on a question to add it to the student response input
3. Click "Send" to submit the response. The question in the question bank should show that it has been used.
4. Click on a question to add it to the student response input
5. Delete the question from the student response input. Type something else and submit the response. The question should not show that it has been used.
6. Leave the step
7. Go back to the Peer Chat step. The questions should show which ones have been used.

Make sure the old Peer Chat Question Bank still works as before for teacher and student.

Closes #1413

### CODE REVIEW

**Review by breity — CHANGES_REQUESTED** [review]
Functionality looks great. I made some style changes and added some text for screen readers.

I found one minor bug. To reproduce:
- Click on one of the question buttons to add to response input
- Delete the text in the response input
- Click on the same question again

The question won't be added to the response input.

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Looking good. I have some comments:
- It would be nice to explain what the "Enable Student Click To Add" option is, or make it more obvious, like "Student can click on items to add to peer chat"
- Are we assuming that they always start by choosing a question bank item first? If I write something and then choose a question, it overwrites what I typed. Should the questions append instead of overwriting? 
- Grading view is not working for me. I see this and an error in the console:
Screenshot 2023-09-13 at 2 54 04 PM
Screenshot 2023-09-13 at 2 54 11 PM

- Rename ```questionClicked()``` to ```addQuestionToChat()``` or something similar?

**Comment by geoffreykwan** [comment]
I changed the component content field from "clickToAddEnabled" to "clickToUseEnabled" so you'll need to check the "Student can click questions to use in Peer Chat" check box in the Peer Chat Question Bank authoring to enable the feature again.

File: src/assets/wise5/components/peerChat/peer-chat-question-bank/question-bank-helper.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export function getQuestionIdsUsed(componentStates: any[], workgroupId: number): string[] {
+  return componentStates
+    .filter((componentState) => componentState.workgroupId === workgroupId)
+    .filter((componentState) => componentState.studentData.questionId != null)
```
Comment (hirokiterashima) [inline src/assets/wise5/components/peerChat/peer-chat-question-bank/question-bank-helper.ts]: Combine into one filter to remove extra loop?

File: src/assets/wise5/components/peerChat/peer-chat-student/peer-chat-student.component.ts [hunk 2]
```diff
@@ -270,4 +283,23 @@ export class PeerChatStudentComponent extends ComponentStudent {
   onDynamicPromptChanged(feedbackRule: FeedbackRule): void {
     this.dynamicPrompt = feedbackRule;
   }
+
+  protected useQuestion(question: Question): void {
+    if (
+      this.response === '' ||
+      confirm(
+        $localize`Are you sure you want to put this question into your input box? The question will overwrite the current text in your input box.`
```
Comment (hirokiterashima) [inline src/assets/wise5/components/peerChat/peer-chat-student/peer-chat-student.component.ts]: Should we use the term "question" here, or use something more generic, like "item"? Some authors may rename "Question Bank" to "Prompt Bank".

**Review by breity — APPROVED** [review]
Looks good. 👍 

I updated the language slightly for the label for the authoring tool setting ("click" -> "select") and the student overwrite confirmation for clarity/brevity.

File: src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts [hunk 3]
```diff
@@ -31,4 +35,14 @@ export class PeerChatMessageInputComponent implements OnInit {
     this.messageText = '';
     this.isSubmitEnabled = false;
   }
+
+  onFocus(event: any): void {
+    this.placeCursorAtEndOfMessageText(event);
+  }
+
+  placeCursorAtEndOfMessageText(event: any): void {
```
Comment (hirokiterashima) [inline src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts]: private?

File: src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts [hunk 4]
```diff
@@ -31,4 +35,14 @@ export class PeerChatMessageInputComponent implements OnInit {
     this.messageText = '';
     this.isSubmitEnabled = false;
   }
+
+  onFocus(event: any): void {
```
Comment (hirokiterashima) [inline src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts]: protected?

File: src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
 })
 export class PeerChatMessageInputComponent implements OnInit {
-  @Output('onSubmit')
-  submit: EventEmitter<string> = new EventEmitter<string>();
-
   isSubmitEnabled: boolean = false;
-  messageText: string;
+  @Input() messageText: string = '';
+  @Output() responseChangedEvent: EventEmitter<string> = new EventEmitter<string>();
+  @Output('onSubmit') submit: EventEmitter<string> = new EventEmitter<string>();
```
Comment (hirokiterashima) [inline src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts:11]: add private/protected as appropriate?

File: src/app/authoring-tool/edit-question-bank/edit-question-bank.component.ts [hunk 6]
```diff
@@ -11,16 +12,29 @@ import { TeacherProjectService } from '../../../assets/wise5/services/teacherPro
 export class EditQuestionBankComponent implements OnInit {
   allowedReferenceComponentTypes: string[] = ['OpenResponse'];
   @Input() componentContent: any;
+  inputChanged: Subject<string> = new Subject<string>();
+  subscriptions: Subscription = new Subscription();
```
Comment (hirokiterashima) [inline src/app/authoring-tool/edit-question-bank/edit-question-bank.component.ts]: Add private/protected as appropriate?

File: src/app/authoring-tool/edit-question-bank-rules/edit-question-bank-rules.component.spec.ts [hunk 7]
```diff
@@ -42,12 +43,35 @@ describe('EditQuestionBankRulesComponent', () => {
...
 function addNewFeedbackToRule() {
   describe('addNewFeedbackToRule()', () => {
+    addNewFeedbackToRuleVersion1();
+    addNewFeedbackToRuleVersion2();
+  });
+}
+
+function addNewFeedbackToRuleVersion1() {
+  describe('using question bank content version 1', () => {
+    it('should add new question to rule', () => {
```
Comment (hirokiterashima) [inline src/app/authoring-tool/edit-question-bank-rules/edit-question-bank-rules.component.spec.ts]: "adds new question to rule"? Remove "should".

**Review by hirokiterashima — APPROVED** [review]
Looks good. I added minor code improvement suggestions to consider before merging.

File: src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts (hunk 5 above)
Comment (geoffreykwan) [reply src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts:11]: @Input() and @Output() variables should be public right?

Comment (hirokiterashima) [reply src/assets/wise5/components/peerChat/peer-chat-message-input/peer-chat-message-input.component.ts:11]: Ah, right- in this case, it's public. You can ignore my comment.

I just wanted to make sure that we add modifiers to all new field variables, where appropriate.

**Pull request merged**